File: apps/scala/src/main/scala/algorithm/Trees.scala
package algorithm

import scala.collection.mutable.ArrayBuffer

/** 第9章 木構造 */

class BSTEmptyException extends RuntimeException("木は空です")

/** 二分探索木 */
class BinarySearchTree[T: Ordering]:
  import Ordering.Implicits.*

  private class Node(var key: T, var left: Node | Null = null, var right: Node | Null = null)

  private var root: Node | Null = null
  private var _size = 0

  def size: Int    = _size
  def isEmpty: Boolean = root == null
  def contains(key: T): Boolean = search(key) != null

  private def search(key: T): Node | Null =
    var ptr = root
    while ptr != null do
      val cmp = summon[Ordering[T]].compare(key, ptr.key)
      if cmp == 0 then return ptr
      ptr = if cmp < 0 then ptr.left else ptr.right
    null

  def insert(key: T): Unit =
    if root == null then { root = Node(key); _size += 1; return }
    var ptr = root
    while true do
      val cmp = summon[Ordering[T]].compare(key, ptr.key)
      if cmp == 0 then return // 重複は無視
      if cmp < 0 then
        if ptr.left == null then { ptr.left = Node(key); _size += 1; return }
        ptr = ptr.left
      else
        if ptr.right == null then { ptr.right = Node(key); _size += 1; return }
        ptr = ptr.right

  def delete(key: T): Unit =
    var parent: Node | Null = null
    var ptr = root
    var isLeftChild = false
    while ptr != null do
      val cmp = summon[Ordering[T]].compare(key, ptr.key)
      if cmp == 0 then
        ptr = null // break
      else
        parent = ptr
        if cmp < 0 then { isLeftChild = true;  ptr = ptr.left  }
        else             { isLeftChild = false; ptr = ptr.right }

    val node = search(key)
    if node == null then return
    _size -= 1

    def replace(child: Node | Null): Unit =
      if parent == null then root = child
      else if isLeftChild then parent.left = child
      else parent.right = child

    if node.left == null && node.right == null then replace(null)
    else if node.right == null then replace(node.left)
    else if node.left == null then replace(node.right)
    else
      var succParent = node
      var succ = node.right
      while succ.left != null do
        succParent = succ; succ = succ.left
      node.key = succ.key
      if succParent == node then succParent.right = succ.right
      else succParent.left = succ.right

  def inOrder(): List[T] =
    val result = ArrayBuffer[T]()
    def traverse(n: Node | Null): Unit =
      if n != null then
        traverse(n.left); result += n.key; traverse(n.right)
    traverse(root)
    result.toList

  def preOrder(): List[T] =
    val result = ArrayBuffer[T]()
    def traverse(n: Node | Null): Unit =
      if n != null then
        result += n.key; traverse(n.left); traverse(n.right)
    traverse(root)
    result.toList

  def postOrder(): List[T] =
    val result = ArrayBuffer[T]()
    def traverse(n: Node | Null): Unit =
      if n != null then
        traverse(n.left); traverse(n.right); result += n.key
    traverse(root)
    result.toList

File: apps/fsharp/Algorithm/Trees.fs
module Trees

// ─── 二分探索木 ────────────────────────────────────────────────

exception BSTEmptyException of string

[<AllowNullLiteral>]
type private BSTNode<'T>(key: 'T) =
    let mutable _key = key
    let mutable _left: BSTNode<'T> = null
    let mutable _right: BSTNode<'T> = null
    member _.Key with get() = _key and set v = _key <- v
    member _.Left with get() = _left and set v = _left <- v
    member _.Right with get() = _right and set v = _right <- v

type BinarySearchTree<'T when 'T : comparison>() =
    let mutable root: BSTNode<'T> = null
    let mutable size = 0

    member _.Size() = size
    member _.IsEmpty() = root = null

    member _.Contains(key: 'T) =
        let mutable ptr = root
        let mutable found = false
        while ptr <> null && not found do
            if key = ptr.Key then found <- true
            elif key < ptr.Key then ptr <- ptr.Left
            else ptr <- ptr.Right
        found

    member _.Insert(key: 'T) =
        if root = null then
            root <- BSTNode<'T>(key)
            size <- size + 1
        else
            let mutable ptr = root
            let mutable inserted = false
            while not inserted do
                if key = ptr.Key then
                    inserted <- true  // 重複は無視
                elif key < ptr.Key then
                    if ptr.Left = null then
                        ptr.Left <- BSTNode<'T>(key)
                        size <- size + 1
                        inserted <- true
                    else
                        ptr <- ptr.Left
                else
                    if ptr.Right = null then
                        ptr.Right <- BSTNode<'T>(key)
                        size <- size + 1
                        inserted <- true
                    else
                        ptr <- ptr.Right

    member this.Delete(key: 'T) =
        let mutable parent: BSTNode<'T> = null
        let mutable ptr = root
        let mutable isLeft = false
        let mutable found = false
        while ptr <> null && not found do
            if key = ptr.Key then found <- true
            else
                parent <- ptr
                if key < ptr.Key then
                    isLeft <- true
                    ptr <- ptr.Left
                else
                    isLeft <- false
                    ptr <- ptr.Right
        if ptr = null then ()
        else
            size <- size - 1
            let replace (n: BSTNode<'T>) =
                if parent = null then root <- n
                elif isLeft then parent.Left <- n
                else parent.Right <- n
            if ptr.Left = null && ptr.Right = null then
                replace null
            elif ptr.Right = null then
                replace ptr.Left
            elif ptr.Left = null then
                replace ptr.Right
            else
                // 中順後継（右部分木の最小値）を見つける
                let mutable succParent = ptr
                let mutable succ = ptr.Right
                while succ.Left <> null do
                    succParent <- succ
                    succ <- succ.Left
                ptr.Key <- succ.Key
                if succParent = ptr then succParent.Right <- succ.Right
                else succParent.Left <- succ.Right
                size <- size + 1

    member _.Min() =
        if root = null then raise (BSTEmptyException "木は空です")
        let mutable p = root
        while p.Left <> null do p <- p.Left
        p.Key

    member _.Max() =
        if root = null then raise (BSTEmptyException "木は空です")
        let mutable p = root
        while p.Right <> null do p <- p.Right
        p.Key

    member _.InOrder() =
        let result = System.Collections.Generic.List<'T>()
        let rec traverse (n: BSTNode<'T>) =
            if n <> null then
                traverse n.Left
                result.Add(n.Key)
                traverse n.Right
        traverse root
        result |> Seq.toList

    member _.PreOrder() =
        let result = System.Collections.Generic.List<'T>()
        let rec traverse (n: BSTNode<'T>) =
            if n <> null then
                result.Add(n.Key)
                traverse n.Left
                traverse n.Right
        traverse root
        result |> Seq.toList

    member _.PostOrder() =
        let result = System.Collections.Generic.List<'T>()
        let rec traverse (n: BSTNode<'T>) =
            if n <> null then
                traverse n.Left
                traverse n.Right
                result.Add(n.Key)
        traverse root
        result |> Seq.toList
